packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -1510,490 +1528,126 @@ class _ActionSheetMainSheetState extends State<_ActionSheetMainSheet> {
-
-    if (height.isFinite) {
-      return height;
-    }
-    return 0.0;
-  }
-
-  @override
-  Size computeDryLayout(BoxConstraints constraints) {
-    return _performLayout(
-      constraints: constraints,
-      layoutChild: ChildLayoutHelper.dryLayoutChild,
-    ).size;
-  }
-
-  @override
-  void performLayout() {
-    final _AlertDialogSizes dialogSizes = _performLayout(
-      constraints: constraints,
-      layoutChild: ChildLayoutHelper.layoutChild,
-    );
-    size = dialogSizes.size;
-
-    // Set the position of the actions box to sit at the bottom of the dialog.
-    // The content box defaults to the top left, which is where we want it.
-    assert(actionsSection!.parentData is BoxParentData);
-    final BoxParentData actionParentData = actionsSection!.parentData! as BoxParentData;
-    actionParentData.offset = Offset(0.0, dialogSizes.contentHeight + dialogSizes.dividerThickness);
-  }
-
-  _AlertDialogSizes _performLayout({required BoxConstraints constraints, required ChildLayouter layoutChild}) {
-    return isInAccessibilityMode
-        ? performAccessibilityLayout(
-            constraints: constraints,
-            layoutChild: layoutChild,
-          ) : performRegularLayout(
-            constraints: constraints,
-            layoutChild: layoutChild,
-          );
-  }
-
-  // When not in accessibility mode, an alert dialog might reduce the space
-  // for buttons to just over 1 button's height to make room for the content
-  // section.
-  _AlertDialogSizes performRegularLayout({required BoxConstraints constraints, required ChildLayouter layoutChild}) {
-    final bool hasDivider = contentSection!.getMaxIntrinsicHeight(getMaxIntrinsicWidth(0)) > 0.0
-        && actionsSection!.getMaxIntrinsicHeight(getMaxIntrinsicWidth(0)) > 0.0;
-    final double dividerThickness = hasDivider ? _dividerThickness : 0.0;
-
-    final double minActionsHeight = actionsSection!.getMinIntrinsicHeight(getMaxIntrinsicWidth(0));
-
-    final Size contentSize = layoutChild(
-      contentSection!,
-      constraints.deflate(EdgeInsets.only(bottom: minActionsHeight + dividerThickness)),
-    );
-
-    final Size actionsSize = layoutChild(
-      actionsSection!,
-      constraints.deflate(EdgeInsets.only(top: contentSize.height + dividerThickness)),
-    );
-
-    final double dialogHeight = contentSize.height + dividerThickness + actionsSize.height;
+  double _topOverscroll = 0;
+  double _bottomOverscroll = 0;
 
-    return _AlertDialogSizes(
-      size: constraints.constrain(Size(_dialogWidth, dialogHeight)),
-      contentHeight: contentSize.height,
-      dividerThickness: dividerThickness,
+  // Fills the overscroll area at the top and bottom of the sheet. This is
+  // necessary because the action section's background is rendered by the
+  // buttons, so that a button's background can be _replaced_ by a different
+  // color when the button is pressed.
+  Widget _buildOverscroll(Color backgroundColor) {
```

Comment:
Absolutely! I planned to do this in a future PR to avoid making this PR even harder to understand, but now that you've noticed it I can refactor it right now.